deploy/s3.sh

```diff
@@ -0,0 +1,189 @@
+    return 1
+  fi
+
+  if [ -z "$S3_REGION" ]; then
+    S3_REGION="us-east-1"
+  fi
+
+  # Save s3 options if it's succesful (First run case)
+  _saveaccountconf S3_BUCKET "$S3_BUCKET"
+  _saveaccountconf S3_REGION "$S3_REGION"
+  _saveaccountconf AWS_PROFILE "$AWS_PROFILE"
+
+  _debug _cdomain "$_cdomain"
+  _debug _ckey "$_ckey"
+  _debug _ccert "$_ccert"
+  _debug _cca "$_cca"
+  _debug _cfullchain "$_cfullchain"
+  _debug S3_BUCKET "$S3_BUCKET"
+  _debug AWS_PROFILE "$AWS_PROFILE"
+  _secure_debug AWS_ACCESS_KEY_ID "$AWS_ACCESS_KEY_ID"
+  _secure_debug AWS_SECRET_ACCESS_KEY "$AWS_SECRET_ACCESS_KEY"
+  
+  # REMOVE BEFORE COMMIT, ONLY FOR DEBUGGING
+  _aws_cli_installed=1
+
+  _info "Deploying certificate to s3 bucket: $S3_BUCKET in $S3_REGION"
+  
+  if [ "$_aws_cli_installed" -eq "0" ]; then
+    _debug "deploying with curl method"
+  else
+    _debug "deploying with aws cli method"
+  fi
+
+  # private
+  _deploy_to_bucket "$_ckey" "$_cdomain/$_cdomain.key"
+  # public
+  _deploy_to_bucket "$_ccert" "$_cdomain/$_cdomain.cer"
+  # ca
+  _deploy_to_bucket "$_cca" "$_cdomain/ca.cer"
+  # fullchain
+  _deploy_to_bucket "$_cfullchain" "$_cdomain/fullchain.cer"
+
+  return 0
+
+}
+
+####################  Private functions below ##################################
+
+_deploy_to_bucket() {
+  if [ "$_aws_cli_installed" -eq "0" ]; then
+    _deploy_with_curl "$1" "$2"
+  else
+    _deploy_with_awscli "$1" "$2"
+  fi
+}
+
+_deploy_with_awscli() {
+  file="$1"
+  prefix="$2"
+  aws s3 cp "$file" s3://"$S3_BUCKET"/"$prefix" --region "$S3_REGION" --profile "$AWS_PROFILE"
+}
+
+_deploy_with_curl() {
+
+  file="${1}"
+  bucket="${S3_BUCKET}"
+  prefix="${2}"
+  region="${S3_REGION}"
+  acl="private"
+  timestamp="$(date -u "+%Y-%m-%d %H:%M:%S")"
+  signed_headers="date;host;x-amz-acl;x-amz-content-sha256;x-amz-date"
+
+  if [ "$(uname)" = "Darwin" ]; then
+    iso_timestamp=$(date -ujf "%Y-%m-%d %H:%M:%S" "${timestamp}" "+%Y%m%dT%H%M%SZ")
+    date_scope=$(date -ujf "%Y-%m-%d %H:%M:%S" "${timestamp}" "+%Y%m%d")
+    date_header=$(date -ujf "%Y-%m-%d %H:%M:%S" "${timestamp}" "+%a, %d %h %Y %T %Z")
+  else
+    iso_timestamp=$(date -ud "${timestamp}" "+%Y%m%dT%H%M%SZ")
+    date_scope=$(date -ud "${timestamp}" "+%Y%m%d")
+    date_header=$(date -ud "${timestamp}" "+%a, %d %h %Y %T %Z")
+  fi
+
+  _info "Uploading $S3_BUCKET/$prefix"
+
+  export _H1
+  export _H2
+  export _H3
+  export _H4
+  export _H5
+  
+  _H1="Authorization: AWS4-HMAC-SHA256 Credential=${AWS_ACCESS_KEY_ID}/${date_scope}/${region}/s3/aws4_request,SignedHeaders=${signed_headers},Signature=$(_signature)"
+  _H2="Date:${date_header}"
+  _H3="x-amz-acl:${acl}"
+  _H4="x-amz-content-sha256:$(_payload_hash)"
+  _H5="x-amz-date:${iso_timestamp}"
+
+  _debug2 "$(_post "${file}" "https://$bucket.s3.$region.amazonaws.com/$prefix")"
+}
+
+####################  Private functions below ##################################
+
+_payload_hash() {
```

Comment:
It makes the script more readable and less repeatable.
I've pushed a change that uses builtin acme.sh function for sha256 digest instead.